### ISSUE TITLE
Namespace for experimental operators

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#209]
Follow up from #205:

>> I don't want to create & maintain another whole library for just one method in the async category. Perhaps a `MoreLinq.Sandbox` would be better for all ideas that are cooking and unsupported?

> This sounds like a good idea, but orthogonal to what I proposed.

Coming back to this, how about introducing a new namespace called `MoreLinq.Labs` in the current, one and only, assembly for experimental ideas with the disclaimer & understanding that anything in there is work-in-progress and could be completely removed in the future. Anyone using something in there should be prepared to deal with breaking changes. If an operator stabilizes over time and proves to be popular then it can be promoted to the `MoreLinq` namespace. The benefits of this approach would be:

- Use existing assembly & NuGet packages instead of maintaining a new and separate one
- Have one release to worry about
- Increase the release cadence of the project by shipping possibly unstable work

Risks? Confusion & some disgruntled users who took the happy path and ignored the writing on the wall.

**Comment by atifaziz** [comment]
Namespace proposals:

`MoreLinq.Edge`: Stuff that's really slated for a future release but shipping part of the current release train. It is potentially unstable & not battle-tested.

`MoreLinq.Labs`: Experimental stuff that may not only be unstable & untested but also may never become part of MoreLINQ proper (therefore risks being completely removed). The reason to ever release this stuff is to summon feedback from wider field usage of existing MoreLINQ users.

**Comment by fsateler** [comment]
There is a smallish problem with including in the same assembly. Visual Studio (I'm using 2015 here) will offer to include the `using` statement necessary for an extension method with the `Ctrl+.` shortcut. This increases the chance someone uses this namespace without fully understanding the consequences.

I looked for an attribute that would generate a compiler warning like `ObsoleteAttribute`, but I couldn't find any. Would it be too annoying/confusing to use `ObsoleteAttribute`?

Another mechanism would be to be more explicit in the namespace name: `MoreLinq.Experimental`.

BTW, how could one get feedback on usage of the experimental methods? It seems to me this is an important part of the point of the exercise, isn't it?

**Comment by atifaziz** [comment]
> Would it be too annoying/confusing to use `ObsoleteAttribute`?

Definitely.

I can live with `MoreLinq.Experimental`. It seems like a good compromise. Not everyone may be current with hipster terms like `Labs`. 🤖 

Yes, the exercise is to deploy & test in the field & solicit feedback. The mechanism to report problems or make fixes would be the usual business of opening an issue or submitting a PR. Why should it be any different? I'm guessing your popping the question because you were imagining something else?

> chance someone uses this namespace without fully understanding the consequences.

So `Experimental` in the namespace should ring alarm bells. And if that's not clear enough then we can go with something really ugly like `MoreLinq.ThereBeDragons`. 🐉 I don't want to let IntelliSense or tooling get in the way of field testing and getting some interesting ideas into the hands of the users. The benefits of the latter are just too many and one of the risks I listed early on was:

> some disgruntled users who took the happy path and ignored the writing on the wall.

The best we can do here is make it clear in the README, NuGet package description and put an ALL CAPS warning in the `<summary>` part of a method's documentation (not just as a passing phrase in the `<remarks>` section). If that won't stop 'em then I'm afraid nothing will. :)

**Comment by JamesFaix** [comment]
MoreLinq.Beta?

**Comment by atifaziz** [comment]
@JamesFaix `MoreLinq.Edge` would be alpha or beta. We don't want to start moving things between namespaces as they stabilise, like when they go from alpha quality to beta. If they change namespace, it should be once. I'm trying to avoid what happened with 2.0, where the number of new operators was large and took long to review. Most operators work independent of each other and those that need more time and attention (perhaps due to lack of time in the first place) shouldn't hold back from releasing others that have fully cooked.

**Comment by JamesFaix** [comment]
I can see how `.Beta` might be too specific.

**Comment by atifaziz** [comment]
I've submitted PR #397 for this based on the following conclusions so far from the discussions:

- Experiments will go in the same assembly & package as the main library
- Naturally, experiments will be in their namespace; explicit opt-in
- ALL CAPS warnings in documentation (i.e. also IntelliSense) and README
- Can be removed in any future major or _minor_ version, and without notice
- We won't care about conflicts (#214)

All of the above points are in light of making experimentation as low maintenance as possible.

If I have jumped the gun on this and there are strong objections than I'm happy to close #397 and we can keep on talking but I thought it might be good to make some progress here. As I also [remarked](https://github.com/morelinq/MoreLINQ/pull/397#issue-171799408) in #397, if this turns out to be a very bad idea, it can be scratched in any minor release. Should someone come rely on any functionality, they have the option to import the source code of an operator into their project.
